lektor/watcher.py

```diff
@@ -2,59 +2,86 @@
 import queue
 import time
 
+import click
 from watchdog.events import DirModifiedEvent
 from watchdog.events import FileMovedEvent
 from watchdog.events import FileSystemEventHandler
 from watchdog.observers import Observer
+from watchdog.observers.polling import PollingObserver
 
 from lektor.utils import get_cache_dir
 
-# Alias this as this can be called during interpreter shutdown
-_Empty = queue.Empty
-
 
 class EventHandler(FileSystemEventHandler):
-    def __init__(self, callback=None):
-        if callback is not None:
-            self.queue = None
-            self.callback = callback
-        else:
-            self.queue = queue.Queue()
-            self.callback = self.queue.put
+    def __init__(self):
+        self.queue = queue.Queue()
 
     def on_any_event(self, event):
         if not isinstance(event, DirModifiedEvent):
-            path = (
-                event.dest_path if isinstance(event, FileMovedEvent) else event.src_path
-            )
-            item = (time.time(), event.event_type, path)
-            if self.queue is not None:
-                self.queue.put(item)
+            if isinstance(event, FileMovedEvent):
+                path = event.dest_path
             else:
-                self.callback(*item)
+                path = event.src_path
+            self.queue.put((time.time(), event.event_type, path))
 
 
 class BasicWatcher:
-    def __init__(self, paths, callback=None):
-        self.event_handler = EventHandler(callback=callback)
-        self.observer = Observer()
-        for path in paths:
-            self.observer.schedule(self.event_handler, path, recursive=True)
-        self.observer.setDaemon(True)
+    def __init__(self, paths, observer_classes=(Observer, PollingObserver)):
+        self.event_handler = EventHandler()
+        self.paths = paths
+        self.observer_classes = observer_classes
+        self.observer = None
+
+    def start(self):
+        untried = set(self.observer_classes)
+        for observer_class in self.observer_classes:
+            try:
+                self._start_observer(observer_class)
```

Comment:
There should be a return below this line, the code currently starts all the observers in sequence

Reply:
Well, that's embarrassing.  Thank you.

Reply:
I've now added a check to prevent starting more than one observer.  (It also guards against someone calling `BasicWatcher.start()` more than once.)